pkg/storageincentives/redistributioncontract/redistribution.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2022 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package redistributioncontract
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethersphere/bee/pkg/log"
+	"github.com/ethersphere/bee/pkg/sctx"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/transaction"
+)
+
+var (
+	redistributionContractABI = parseABI(redistributionABIv0_0_0)
+)
+
+const loggerName = "redistributionContract"
+
+type Interface interface {
+	ReserveSalt(context.Context) ([]byte, error)
+	IsPlaying(context.Context, uint8) (bool, error)
+	IsWinner(context.Context) (bool, error)
+	Claim(context.Context) error
+	Commit(context.Context, []byte) error
+	Reveal(context.Context, uint8, []byte, []byte) error
+}
+
+type Service struct {
+	overlay                   swarm.Address
+	logger                    log.Logger
+	txService                 transaction.Service
+	incentivesContractAddress common.Address
+}
+
+func New(
+	overlay swarm.Address,
+	logger log.Logger,
+	txService transaction.Service,
+	incentivesContractAddress common.Address,
+) *Service {
+
+	s := &Service{
+		overlay:                   overlay,
+		logger:                    logger.WithName(loggerName).Register(),
+		txService:                 txService,
+		incentivesContractAddress: incentivesContractAddress,
+	}
+	return s
+}
+
+func (s *Service) IsPlaying(ctx context.Context, depth uint8) (bool, error) {
+	callData, err := redistributionContractABI.Pack("isParticipatingInUpcomingRound", common.BytesToHash(s.overlay.Bytes()), depth)
+	if err != nil {
+		return false, err
+	}
+
+	result, err := s.callTx(ctx, callData)
+	if err != nil {
+		return false, err
+	}
+
+	results, err := redistributionContractABI.Unpack("isParticipatingInUpcomingRound", result)
+	if err != nil {
+		return false, err
+	}
+
+	return results[0].(bool), nil
+}
+
+func (s *Service) IsWinner(ctx context.Context) (isWinner bool, err error) {
+	callData, err := redistributionContractABI.Pack("isWinner", common.BytesToHash(s.overlay.Bytes()))
+	if err != nil {
+		return false, err
+	}
+
+	result, err := s.callTx(ctx, callData)
+	if err != nil {
+		return false, err
+	}
+
+	results, err := redistributionContractABI.Unpack("isWinner", result)
+	if err != nil {
+		return false, err
+	}
+
+	return results[0].(bool), nil
+}
+
+func (s *Service) Claim(ctx context.Context) error {
+	callData, err := redistributionContractABI.Pack("claim")
+	if err != nil {
+		return err
+	}
+	request := &transaction.TxRequest{
+		To:          &s.incentivesContractAddress,
+		Data:        callData,
+		GasPrice:    sctx.GetGasPrice(ctx),
+		GasLimit:    sctx.GetGasLimitWithDefault(ctx, 9_000_000),
```

Comment:
can this also bee 3_000_000

Reply:
i discussed with Abel today he said that its better to have 9_000_000 for claim as it may use more gas.